haystack_experimental/core/pipeline/pipeline.py

```diff
@@ -31,117 +37,94 @@ class Pipeline(PipelineBase):
-        # might not provide these defaults for components with inputs defined dynamically upon component initialization
-        component_inputs = self._add_missing_input_defaults(component_inputs, component["input_sockets"])
+        :returns:
+            The output of the Component.
+        """
+        instance: Component = component["instance"]
 
         # Deserialize the inputs if they are passed in resume state
         # this check will prevent other inputs generated at runtime from being deserialized
-        if self.resume_state and component_name in self.resume_state["pipeline_state"]["inputs"].keys():
-            for key, value in component_inputs.items():
-                component_inputs[key] = Pipeline._deserialize_component_input(value)
-
-        if breakpoints and not self.resume_state:
-            # add component_inputs to inputs
-            breakpoint_inputs = deepcopy(inputs)
-            # we deepcopy the component_inputs to avoid modifying the original inputs
-            breakpoint_inputs[component_name] = deepcopy(Pipeline._remove_unserializable_data(component_inputs))
-
-            # TODO: Find a better way to handle this.
-            # Using to_dict() here strips away class types like ChatMessage,
-            # which makes deserialization of the params problematic. On the other hand, using __dict__ relies on
-            # class variables being stored with the same names as their constructor parameters.
-
-            # We use copy instead of deepcopy to avoid issues with unpickleable objects like RLock
-            params = copy(component["instance"].__dict__)
-
-            # this is needed as the template param is stored as _template_string in the component's __dict__
-            if "_template_string" in params:
-                params["template"] = params["_template_string"]
-            breakpoint_inputs[component_name]["init_parameters"] = params
-            self._check_breakpoints(breakpoints, component_name, component_visits, breakpoint_inputs)
+        if resume_state and component_name in resume_state["pipeline_state"]["inputs"].keys():
+            for key, value in inputs.items():
+                inputs[key] = deserialize_component_input(value)
+
+        if breakpoints and not resume_state:
+            state_inputs_serialised = remove_unserializable_data(deepcopy(state_inputs))
+            # inject the inputs into the state_inputs so we can this component init params in the JSON state
+            state_inputs_serialised[component_name] = remove_unserializable_data(deepcopy(inputs))
+
+            Pipeline._check_breakpoints(
+                breakpoints=breakpoints,
+                component_name=component_name,
+                component_visits=component_visits,
+                inputs=state_inputs_serialised,
+                debug_path=debug_path,
+                original_input_data=original_input_data,
+                ordered_component_names=ordered_component_names
+            )
```

Comment:
As a conceptual question, why does this need to live in `_run_component`? I quickly pulled and refactored this locally and was able to extract all of this code to be right before `self._run_component` is used.

Reply:
The tests didn't need any changes with the refactor I had in mind. I'll go ahead and push it since I think it's easy to reverse in case you need to in the future.

Reply:
very cool! thanks 👍🏽 I quickly ran it to see if there were any changes on JSON state output - all good 👍🏽